### ISSUE TITLE
Fix exec::exec_va__bool

### ISSUE THREAD

**Issue by jarctan** [jarctan/vache#9]
Why:
- error: "expected Borrowed, found Moved at 8:17"
- because there are actually two elements starting from that span and `Moved` takes precedence over `Borrowed` in that case

How to fix:
Revamp the .vat files to at least include start and end of span to avoid any ambiguity when designating elements and their modes

**Comment by jarctan** [comment]
Changed the .vat file to include start and end of start to remove any ambiguity in 75ec26050077cbaa032ccf750cc5b0feb50e74ad
However, this did not fix the issue

**Comment by jarctan** [comment]
The issue is actually about when should we move or not

Previously, we would move only if it is the last occurrence of a variable (and then all loans valid at that last use are invalidated).

In that case, the `bool` test does not pass, since here:
```
if a > b && a > 0 {
```
The last `a` is moved, but this invalidates the first `a` (which only borrowed and is still active at that point), which must be cloned. Yet, we can see that the best solution would be to borrow them both. So, we could say to move only if there is no (which would get invalidated anyway).

This looks good but triggers a new problem in the `fibo` function. Now, both occurrences of `i` in this snippet are going to be `Borrowed`:

```
fibo(i - 1) + fibo(i - 2)
```

In particular, neither `fibo(i - 1)` nor `fibo(i - 2)` will be able to own `i`. But we could see that it would be better to clone `i` in `i - 1` and move `i` in `fibo(i - 1)`, so that there would be only one clone. But for that optimization, we must know that `fibo` is never going to return use the borrowed variable (it will always try to own it), and own reasoning would be false if we had `fibo(i - 0)` and `fibo(i - 1)`, since in the first case `i` will NOT be cloned and the borrow on `i` will effectively linger till `fibo(i -1)`, which could not move anyway.

A more complex example:

```
while i > 0 {
    n2 = n2 + n1;
    n1 = n2 - n1;
    i = i - 1;
}
```
Here, the computed borrows are not the optimized ones. `n2` in `n2 = n2 + n1` is now moved. Why? Because of the order of invalidations, that will invalidate (for instance) the second `n1`, then the first `n1`, then the first `n2`, and finally the second `n2`. There are other invalidations orders that also lead to suboptimal number of clones (cloning the first `n2`, then the first `n1`, then the second `n2`)

A more detailed example that leads to an suboptimal number of clones:
```
Invalidations: {点23|n1§16%n2, 点27|n1§14%時16, 点22|n2§16%時15, 点26|n2§14%n1}
Invalidated: 点23|n1§16%n2
Could not clone because of [点26|n2§14%n1]
Could not clone because of [点26|n2§14%n1]
Invalidations: {点22|n2§16%時15, 点26|n2§14%n1}
Invalidated: 点22|n2§16%時15
Invalidations: {点26|n2§14%n1}
Invalidated: 点26|n2§14%n1
Invalidations: {}
```

The optimal solution is to clone the first n1, then the second n2. How to know how to resolve invalidations in this order? This is not in the order of the control flow (since in that case, we would invalidate the first `n1` and the first `n2`, which leads to another suboptimal state).

**Comment by jarctan** [comment]
Fixed by 052adfe23455a7ff73f51b37e9a65d2de009d516 (possibly overfitting for that particular example)
